# BONUS EXERCISE   
**Name:** Mariem Memmi  
**Student ID:** 10074879  

## Goals
1. Read two CSV files  
2. Output statistics (**min**, **max**, **count**, **std**, **mean**) for each column in each file
3. Merge both files on `Time` as a common key
4. Train a Linear Regression model with `Energy delta[Wh]` as the target variable  
5. Output the **R² score**
6. Plot a correlation matrix (target + features)
7. Create a scatter plot of `KWH` vs `GHI` with a regression line.

In [2]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
from pathlib import Path

In [3]:
DATA_DIR = Path("data")
OUT_DIR = Path("outputs")
OUT_DIR.mkdir(exist_ok=True)

print("DATA:", DATA_DIR.resolve())
print("OUT :", OUT_DIR.resolve())

DATA: C:\Users\Meriam\Desktop\bonus_points\data
OUT : C:\Users\Meriam\Desktop\bonus_points\outputs


## 1) Load CSV files
- **file1** contains `Time` and `Energy delta[Wh]`
- **file2** contains weather and time-related features

### load 1st csv file

In [6]:
df1 = pd.read_csv(DATA_DIR / "file1.csv", sep=";")
display(df1.head())
print(df1.columns)
print(df1.dtypes)
print(f"\nNaNs in df1:\n{df1.isna().sum()}\n")
print(f"\nDuplicate rows in df1:\n{df1.duplicated().sum()}\n")
print("Rows file1:", len(df1))

,Time,Energy delta[Wh]
0,2017-01-01 00:00:00,0
1,2017-01-01 00:15:00,0
2,2017-01-01 00:30:00,0
3,2017-01-01 00:45:00,0
4,2017-01-01 01:00:00,0


Index(['Time', 'Energy delta[Wh]'], dtype='object')
Time                object
Energy delta[Wh]     int64
dtype: object

NaNs in df1:
Time                0
Energy delta[Wh]    0
dtype: int64


Duplicate rows in df1:
0

Rows file1: 196776


In [7]:
df2 = pd.read_csv(DATA_DIR / "file2.csv")
display(df2.head())
print(df2.columns)
print(df2.dtypes)
print(f"\nNaNs in df2:\n{df2.isna().sum()}\n")
print(f"\nDuplicate rows in df2:\n{df2.duplicated().sum()}\n")
print("Rows file2:", len(df2))

,Time,GHI,temp,pressure,humidity,wind_speed,rain_1h,snow_1h,clouds_all,isSun,sunlightTime,dayLength,SunlightTime/daylength,weather_type,hour,month
0,2021-08-07 19:30:00,0.0,19.8,1005,70,2.0,0.0,0.0,100,0,0,915,0.00,4,19,8
1,2017-01-14 20:45:00,0.0,-1.2,1005,91,3.6,0.0,0.0,28,0,0,480,0.00,2,20,1
2,2018-08-13 06:15:00,75.0,17.3,1010,70,2.3,0.0,0.0,94,1,180,900,0.20,4,6,8
3,2018-09-10 07:00:00,80.7,16.2,1021,69,2.6,0.0,0.0,100,1,165,780,0.21,4,7,9
4,2020-05-07 17:00:00,37.7,12.9,1021,59,3.1,0.0,0.0,71,1,840,930,0.90,3,17,5


Index(['Time', 'GHI', 'temp', 'pressure', 'humidity', 'wind_speed', 'rain_1h',
       'snow_1h', 'clouds_all', 'isSun', 'sunlightTime', 'dayLength',
       'SunlightTime/daylength', 'weather_type', 'hour', 'month'],
      dtype='object')
Time                       object
GHI                       float64
temp                      float64
pressure                    int64
humidity                    int64
wind_speed                float64
rain_1h                   float64
snow_1h                   float64
clouds_all                  int64
isSun                       int64
sunlightTime                int64
dayLength                   int64
SunlightTime/daylength    float64
weather_type                int64
hour                        int64
month                       int64
dtype: object

NaNs in df2:
Time                      0
GHI                       0
temp                      0
pressure                  0
humidity                  0
wind_speed                0
rain_1h               

## 2) Statistics for both files
For each **numeric** column in each file independently, the following statistics are calculated then generated into 2 sepearte csv files (`stats_file1` and `stats_file2`):
- **min**
- **max**
- **count**
- **standard deviation (std)**
- **mean**

In [9]:
# --- stats for file1 columns
stats1 = df1.select_dtypes("number").agg(["min","max","count","std","mean"]).T
print("\n=== file1 stats ===")
print(stats1)

# --- stats for file2 columns
stats2 = df2.select_dtypes("number").agg(["min","max","count","std","mean"]).T
print("\n=== file2 stats ===")
print(stats2)

# --- save stats in two seperate csv files
stats1.to_csv(OUT_DIR / "stats_file1.csv")
stats2.to_csv(OUT_DIR / "stats_file2.csv")



=== file1 stats ===
                  min      max     count          std        mean
Energy delta[Wh]  0.0  50000.0  196776.0  1157.490585  578.024525

=== file2 stats ===
                          min         max     count          std         mean
GHI                       0.0    83500.00  196781.0   253.573447    33.384912
temp                    -16.6    21300.00  196781.0    83.839240    10.176997
pressure                977.0  1030000.00  196781.0  5153.589267  1041.270509
humidity                 22.0    94000.00  196781.0   342.819971    81.424548
wind_speed                0.0     4900.00  196781.0    16.630997     4.017939
rain_1h                   0.0        8.09  196781.0     0.278910     0.066033
snow_1h                   0.0        2.82  196781.0     0.069709     0.007147
clouds_all                0.0    98000.00  196781.0   347.964106    67.461676
isSun                     0.0     1000.00  196781.0     3.225301     0.530112
sunlightTime              0.0   840000.00  196

## 3) Merging the two datasets into one using one common key
- the two datasets **file1** and **file2** are merged into one on a common key which is `Time`.
- the key `Time` must first be converted to datetime because the `Time` column in the 2 files does not have the same format
- inner join ensures that when we merge on `Time`, only the rows whose `Time` values appear in both **file1** and **file2**

In [11]:
# --- time conversion to datetime 
df1["Time"] = pd.to_datetime(df1["Time"], errors="coerce")
df2["Time"] = pd.to_datetime(df2["Time"], errors="coerce")
# --- drop rows where the datetime conversion failed (if ever there are any) 
df1 = df1.dropna(subset=["Time"])
df2 = df2.dropna(subset=["Time"])
# --- merge using inner join
merged_df = pd.merge(df1, df2, on="Time", how="inner").sort_values("Time")
display(merged_df.head())
print("Merged shape:", merged_df.shape)
#  --- save as a csv file
merged_df.to_csv(OUT_DIR / "merged_df.csv", index=False)



,Time,Energy delta[Wh],GHI,temp,pressure,humidity,wind_speed,rain_1h,snow_1h,clouds_all,isSun,sunlightTime,dayLength,SunlightTime/daylength,weather_type,hour,month
0,2017-01-01 00:00:00,0,0.0,1.6,1021,100,4.9,0.0,0.0,100,0,0,450,0.0,4,0,1
1,2017-01-01 00:15:00,0,0.0,1.6,1021,100,4.9,0.0,0.0,100,0,0,450,0.0,4,0,1
2,2017-01-01 00:30:00,0,0.0,1.6,1021,100,4.9,0.0,0.0,100,0,0,450,0.0,4,0,1
3,2017-01-01 00:45:00,0,0.0,1.6,1021,100,4.9,0.0,0.0,100,0,0,450,0.0,4,0,1
4,2017-01-01 01:00:00,0,0.0,1.7,1020,100,5.2,0.0,0.0,100,0,0,450,0.0,4,1,1


Merged shape: (196776, 17)


## 5) Modeling: Linear Regression
Target variable: `Energy delta[Wh]` 
Features: all other columns except of `Time` after proper encoding for `hour`, `month` and `weather_type`

The dataset is split by **time**:
- Train = first 80% rows (oldest data)
- Test  = last 20% rows (newest data)
  
Then model performance is evaluated using **R² score**.


In [13]:
# --- features and target
y = merged_df["Energy delta[Wh]"]
X = merged_df.drop(columns=["Energy delta[Wh]", "Time"]).copy()

# --- cyclical encoding for column "hour" and column "month"
#cyclical encoding used so that the model can understand that "month" and "hour" repeat in a cycle (e.g., hour 23 is close to hour 0, and December(12) is close to January(1)).
X["hour_sin"] = np.sin(2 * np.pi * X["hour"] / 24) # 0..23
X["hour_cos"] = np.cos(2 * np.pi * X["hour"] / 24)
m = X["month"] - 1  # 0..11
X["month_sin"] = np.sin(2*np.pi*m/12)
X["month_cos"] = np.cos(2*np.pi*m/12)
#original "hour" and "month" are dropped because they are replaced with the cyclical versions
X = X.drop(columns=["hour", "month"])

# --- one-hot encode "weather_type" (because "weather_type" is a category)
X = pd.get_dummies(X, columns=["weather_type"], drop_first=True)

# --- time split: first 80% train, last 20% test
split = int(len(X) * 0.8)
X_train, X_test = X.iloc[:split], X.iloc[split:]
y_train, y_test = y.iloc[:split], y.iloc[split:]

# --- train linear regression + R²
model = LinearRegression()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)
model_r2 = r2_score(y_test, y_pred)
print("Model R²    =", model_r2)



Model R²    = 0.6763455912329457


The trained model achieved an R² of  **$\approx$0.68** on the test set, meaning it explains approximately **68%** of the variance in `Energy delta[Wh]` on the test data.